File: leetcode/36.valid-sudoku.py
# @leet start
import itertools

class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        # Validate rows
        for i in range(9):
            seen_set = set()

            for j in range(9):
                val = board[i][j]

                if val != "." and val in seen_set:
                    return False

                seen_set.add(val)

        # Validate columns
        for j in range(9):
            seen_set = set()

            for i in range(9):
                val = board[i][j]

                if val != "." and val in seen_set:
                    return False

                seen_set.add(val)

        # Validate sub-boxes
        def validate_subbox(ul_i: int, ul_j: int) -> bool:
            """ul_i is upper left row, ul_j is upper left column."""
            seen_set = set()

            for i, j in itertools.product(range(ul_i, ul_i + 3), range(ul_j, ul_j + 3)):
                val = board[i][j]

                if val != "." and val in seen_set:
                    return False

                seen_set.add(val)

            return True

        for i, j in itertools.product(range(0, 9, 3), range(0, 9, 3)):
            if not validate_subbox(i, j):
                return False

        return True
# @leet end

File: leetcode/36.valid-sudoku.js
// @leet start
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  const ROWS = 9;
  const COLS = 9;

  const isCellValid = (row, col, seen) => {
    const numStr = board[row][col];

    if (numStr === ".") {
      return true;
    }

    const num = Number(numStr);

    if (num < 1 || num > 9 || seen.has(num)) {
      return false;
    }

    seen.add(num);

    return true;
  };

  for (let row = 0; row < ROWS; row++) {
    const seen = new Set();

    for (let col = 0; col < COLS; col++) {
      if (!isCellValid(row, col, seen)) {
        return false;
      }
    }
  }

  for (let col = 0; col < COLS; col++) {
    const seen = new Set();

    for (let row = 0; row < ROWS; row++) {
      if (!isCellValid(row, col, seen)) {
        return false;
      }
    }
  }

  const subBoxValid = (startRow, startCol) => {
    const seen = new Set();

    for (let i = 0; i < 3; i++) {
      for (let j = 0; j < 3; j++) {
        if (!isCellValid(startRow + i, startCol + j, seen)) {
          return false;
        }
      }
    }

    return true;
  };

  // validate the sub-box i sub-boxes from the left and j from the top
  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      if (!subBoxValid(3 * i, 3 * j)) {
        return false;
      }
    }
  }

  return true;
};
// @leet end
